### ISSUE TITLE
IVY writes metadata into our symlinked monorepo's non-angular library's package.json

### ISSUE THREAD

**Issue by JonWallsten** [angular/angular#33395]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler?

### Is this a regression?

 Can't tell

### Description

When enabling IVY, our symlinked monorepo packages that are imported by the app using IVY, get's metadata written in their package.json by the compiler. I guess the compiler think it's a normal package in node_modules that wouldn't mind this. 

```
#Added as a script
"prepublishOnly": "node --eval \"console.error('ERROR: Trying to publish a package that has been compiled by NGCC. This is not allowed.\\nPlease delete and rebuild the package, without compiling with NGCC, before attempting to publish.\\nNote that NGCC may have been run by importing this package into another project that is being built with Ivy enabled.\\n')\" && exit 1"

#Added as it's own key:
"__processed_by_ivy_ngcc__": {
    "fesm2015": "9.0.0-next.13",
    "fesm5": "9.0.0-next.13",
    "es2015": "9.0.0-next.13",
    "esm2015": "9.0.0-next.13",
    "esm5": "9.0.0-next.13",
    "main": "9.0.0-next.13",
    "module": "9.0.0-next.13"
  }
```

## 🔬 Minimal Reproduction

 https://github.com/JonWallsten/monorepo-repro

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 9.0.0-next.15
Node: 10.16.3
OS: win32 x64
Angular: 9.0.0-next.13
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-next.15
@angular-devkit/build-optimizer   0.900.0-next.15
@angular-devkit/core              9.0.0-next.15 (cli-only)
@angular-devkit/schematics        9.0.0-next.15
@angular/cdk                      9.0.0-next.0
@angular/cli                      9.0.0-next.15
@angular/http                     8.0.0-beta.10
@angular/material                 9.0.0-next.0
@ngtools/webpack                  9.0.0-next.15
@schematics/angular               9.0.0-next.15
@schematics/update                0.900.0-next.15
rxjs                              6.5.3
typescript                        3.5.3
</code></pre>

**Anything else relevant?**

**Comment by petebacondarwin** [comment]
Hi @JonWallsten - for the Angular compiler (ngtsc) to be able to build your applications in ivy mode, we need all its dependencies to be compiled in ivy mode too. We have a second compiler (ngcc) that will re-compiled dependent packages into the ivy mode. Ngcc will identify packages that have been built with view engine and recompile them (in-place effectively).

This is what is happening in your project. This has to happen for all dependencies whether they are in node_modules or not.

One way to avoid this, if the dependency is a local project and not going to be distributed on npm, is to actually compile these packages in ivy mode and then expose the build files to the downstream application rather than the view engine builds. At this stage, if the package is going to be distributed on npm then it should be built in view engine mode (i.e. non-ivy).

**Comment by JonWallsten** [comment]
I'm not following completely, but I get the gist of it. What I don't understand though, is why my non-angular libs (they are not in the example repo) are being compiled at all at this stage. They are already compiled by Webpack/TypeScript in an earlier stage of the build process, as commonjs modules and then we import stuff in the app. They shouldn't have to be touched by angulars compiler, or am I wrong? If they do I guess it would be nice to detect if this is a installed package or a original local package, and don't write stuff in the package.json, would that be possible?

**Comment by alxhub** [comment]
If those packages are Angular packages, they will need to be "updated" from View Engine code to Ivy code.

Another way to think about ngcc is that it's running the Angular compiler on your libraries.

**Comment by JonWallsten** [comment]
@alxhub: One of them are, and I can aceept that it's adding stuff to the package.json in that one. But the other libs are not containing any angular code. They are only guilty of being used by our angular app.

**Comment by marcj** [comment]
I think this is a fundamental design flaw in Ivy's compiler architecture, especially in a local development workflow for linked monorepos or linked libraries in general. We use Lerna for example 

```
packages
├── cli
├── core
├── core-node
├── app
├── electron
├── server
└── website
```

and whenever we build the project the `package.json` of `core` gets changed (because its used by `app`, `website`). It's not clear to me how changing my actual code from a compiler could anything but unpleasant. How is the workflow supposed to be? My Git tools show now all the time changes which I explicitly need to reset before commiting changes. Tools like Jetbrains Webstorm can exclude lines in commits but not the Git CLI, which I use most of the time. I personally find that very frustrating and bad practise.

Beside the bad UX of that another topic is actual a major regression: We use external libraries like https://github.com/marcj/angular-desktop-ui which is a TS-only library. We symlink it locally while development to make the workflow easy. Like so:

```
workspace/
├── application/
├── angular-desktop-ui/
```

where `application/node_modules/angular-desktop-ui` is linked to `../../angular-desktop-ui`.

With Ivy it's not possible anymore to use that setup with more than one consumer where each uses different versions of Angular. Imagine a second application with Angular 9.0.1 that consumers `angular-desktop-ui` (or any other local checkedout and symlinked library), while the first application uses Angular 9.0.0. 


```
workspace/
├── application1/ angular@9.0.0
├── application2/ angular@9.0.1
├── angular-desktop-ui/
```

It asks now everytime to delete the node_modules folder, however after symlinking back `angular-desktop-ui` the same error appears again. Removing the lines manually (super ugly) `__processed_by_ivy_ngcc__` in `angular-desktop-ui/package.json` solves it only for one consumer. When building the other application the same error appears again and you have again to manually remove `__processed_by_ivy_ngcc__`.

The assumption that deleting `node_modules` resets all `package.json` and thus fixes the ngcc issue (as told in the error message `The ngcc compiler has changed since the last ngcc build.`) is wrong in a monorepo or symlinked libraries workflow, which leads not only to horrible UX but also major regressions regarding local development workflows.

So for me there are 2 issues:

1. Modifying `package.json` results in changing my actual local code (not a node_modules/ dependency which I don't care when you change their content) which is a Git repository and thus produces changes. That's a very bad idea for monorepos and alike. At least use please a dedicated file like `.angular.build` which I can put into `.gitignore`. Or better don't change anything in my actual code or its dependencies and put it in an external folder like `~/.angular/build/<project-path>/build.json`.

2. Not being able to use multiple consumers for a symlinked package. This is actually the biggest deal-breaker for me as I have a couple of shared libraries that are used across multiple projects and I'm not able to use Ivy in those setups as it leads to a big headache with that ngcc compiler thing trying to build only one version at a time instead of allowing to coordinate its build artifacts in a way that includes its consumer path/version in its cache keys.

For me it seems Ivy completely ignored the legit workflow of multi-packages symlinked together using either Lerna or other local multi-package workflow patterns. 

What's Angular's approach to fixing that?

**Comment by filipesilva** [comment]
@marcj I'm not really sure how you are building your monorepo in the case you described. I get the impression you're using TS paths to make consuming apps use the library TS sources directly. I think that's the case because you mention the original package.json is modified. 

This is not a use case that we've ever supported in Angular CLI. Instead, we try to setup projects so that consumers use built versions of libraries. If you have libraries generated with Angular CLI, they are also updated to directly compile to Ivy in development workflows.

The symlinked package case is also not very clear to me... symlinked packages have always been a problem because they are usually broken with peer dependencies. Some setups might addressed this of course, but the devil is in the details. The setup you're describing sounds like it's using a single built (or maybe non-built) library linked to several consumers. The same built library should be able to be consumed across different Angular versions, but the same cannot be said of the particular installed version on disk for a given project.

I think you've made a pretty good description of this problem but before we can jump in and describe ways to address these issues, we'd really need to see a example reproduction. This is a very non-trivial case that hinges on how things are built, linked, and consumed. There are setups that might have worked, but not because we explicitly supported them.

**Comment by marcj** [comment]
@filipesilva sure, no problemo. Here is a full prepared and working repo with Lerna, 3 packages, and step-by-step doc on how to reproduce it: https://github.com/marcj/angular-ivy-issue1. This setup here is a very basic and common one in Lerna: Frontend and server share some code which is stored in a `core` package. In this case a simple `User` model. The result is that our `packages/core/package.json` is always changed when building the frontend + we can not publish that core package anymore (thanks to the insert script hook). As already described.

A more complex scenario could be to have a package external of lerna (like described with angular-desktop-ui) and use a solution like [npm-local-development](https://github.com/marcj/npm-local-development) to link it correctly so `devDependencies` are correctly handled. `peerDependencies` were never a issue in my experience. If you use Lerna and that `core` package has own `devDependencies` (which a consumer defined as well) for its unit tests then you need the better linking approach like in npm-local-development as well. 

However, that overall topic with linking external packages or packages in a mono-repo using links is a solved one and works fine (at least if you use a solution like npm-local-development for more complex setups).

> There are setups that might have worked, but not because we explicitly supported them.

Which I think you should consider supporting explicitly. I did dozens of enterprise projects in Europe and my experience is that in TS projects mono-repos and linked packages are pretty common. Breaking this now with Ivy pushes people away from Angular because the only alternative is not to develop on multiple packages that partially use each other locally anymore. I think https://nx.dev/ has the same issue here.

**Comment by coryrylan** [comment]
@marcj I have seen this issue as well in some private repos but also have it in our public Clarity mono repo. We distribute an Angular library as well as a handful of plain TS/CSS libraries as dependencies. I wrote a script as a temporary workaround that runs ngcc during post install removing the added properties.

https://github.com/vmware/clarity/commit/d160f0c9fe5ca89388f054c02b9e18fdd18631b1

**Comment by filipesilva** [comment]
@marcj I followed the instructions on your repo:
```
git clone https://github.com/marcj/angular-ivy-issue1
cd angular-ivy-issue1
npm run install
```

This fails because it needs the `lerna` command somewhere. I installed it using `npm i -g lerna@3.19.0` (the latest) and tried again, and it seems to work:
```
$ npm run install

> root@ install D:\sandbox\angular-ivy-issue1
> lerna bootstrap --nohoist \* --no-ci

lerna notice cli v3.19.0
lerna info Bootstrapping 3 packages
lerna info Installing external dependencies
lerna info Symlinking packages and binaries
lerna success Bootstrapped 3 packages
```

I'm not really familiar with what this command does, but I see that both `packages/server/node_modules/@issue1/core` and `packages/app/server/node_modules/@issue1/core` were created with the TS sources of `packages/core/`. I'm sorry to say but having TS sources in `node_modules` is not something we support. 

Publishing TS sources is generally not safe, since the output of compiling them differs by TS version and by build system setup. If you have a separate Angular library and an app, and you should build the library using whatever build setup you have, then consume the built library from your app. Angular CLI uses `@angular-devkit/build-ngpackagr` to build libraries, but you might build them some other way. If you try to have the app build system consume the library TS files directly, you're no longer building the library and there's no guarantee things would work the same as if you did. In the Angular CLI case, the app and lib build pipelines are very different.

Continuing with your repro I run these commands:
```
cd packages/app
ng build
```

And see the diff you mention in the README.md. `packages/core/package.json` is modified.

This happens because `packages/app/node_modules/@issue1/core/package.json` seem to be a symlink to `packages/core/package.json`. When `ngcc` alters `node_modules` in this sort of setup, it will alter the source files because they are linked.

By design `ngcc` will always alter node_modules. We didn't cover the case of linked TS sources because it was never a supported case to begin with (for the reasons above).

I don't understand why `packages/app/node_modules/@issue1/core/package.json` is altered in the first place though. It's not an angular library. `packages/app/node_modules/zone.js/package.json` wasn't altered and it's imported by the app too. Maybe @petebacondarwin can shed some light on that.

**Comment by marcj** [comment]
@filipesilva thanks for checking. 

> I'm sorry to say but having TS sources in node_modules is not something we support.

This also happens when you have JS sources in `node_modules` that is being imported. Nothing TS related.

> Publishing TS sources is generally not safe, since the output of compiling them differs by TS version and by build system setup.

Unfortunately I can't disagree more here and argue for the exact opposite. It's like saying "don't distribute C++ code because different compiler versions and settings produce different outputs". That behavior to have different outputs based on different settings is exactly what people want when they consume TS only packages. If you as lib author don't support JS consumers it makes absolutely no sense to first compile to JS and then distribute that. I as TS consumer decide what JS version I target with my builds. I decide what features are activated, if three-shaking is used, etc. Its a waste of space and bandwidth when now every lib author prebuilds for many different targets (es5, es6, etc).
Also, if you happen to be the consumer and author at the same time and work on both locally you need to run yet-another-watcher™ just to be able to consume it correctly. That's not only bad UX, but plain not necessary. There is no advantage in distributing JS code in projects that are completely based and only support TS - actually there are only disadvantages if you chose that path. Although if you want to support JS consumers/users as well, there's no question: you have to distribute as JS and share `.d.ts` files along with it.

Despite that, in a monorepo setup you have usually the `tsconfig.json` and TS compiler version sticked in `package-lock.json`, so it's guaranteed to have always the exact same build output. No team on earth uses different TS versions or different compiler settings in the same project. So, even if you ignore the fact that people actually want different outputs for different compiler settings, even hypothetically that argument is not valid.

**Symlinking packages is a inherent property and feature** of NPM and node itself. See `npm link`, see Lerna, see nx.dev, Yarn workspaces, and many more. The only way to work on multiple packages that consume each other at the same time or in a multi-package monorepo is using symlinks. There's literally no other legit workflow available.
Even Angular has options to support symlinks, see `preserveSymlinks` settings in `angular.json`. That's a important setting to make all of this work.
Symlinked packages were supported for ages, that's why TS-only full-stack applications, using Angular + Express server with some common/core packages along with it, all in one repo using Lerna is on the rise. Lerna has meanwhile **2.500.000 installations each month**. That whole workflow using symlinks is a defacto standard and there is no way around that.

Saying the new Angular 9 (concretely only Ivy) doesn't support that workflow anymore because "it was never officially explicitly supported in the first place and now it happens to break because we made the design decision to modify your node_module code" is something I consider a cheap shot. It doesn't have to be supported explicitly to make sure it continues to work. That's also the reason why I consider Ivy with that compiler workflow a huge design flaw. Despite the fact that Angular is now the first framework/library that modifies third-party files in `node_modules` it just breaks that extremely popular workflow, which is without any doubt a huge deal breaker for many projects adopting Ivy. The fix is equally as simple as explaining the obvious design flaw: Just put your meta-data in your own `node_modules/@angular` folder or in users' home directory `~/.angular/ngcc/v9.0.0rc3/build.json`.

**Comment by mlc-mlapis** [comment]
@marcj Thank you for your analysis and arguments. I think that it is something that should be really seriously taken into account by the Angular core team.

**Comment by JonWallsten** [comment]
The main reason for us not compiling/bundling the package is as Marc mentions performance. Why have a Webpack instance running when the consumer app is going to process it later anyways? The reason it's in it's own package in the first place is so we can share code between applications. They can handle the compiling/bundling themselves. 

Also, as I mentioned in a previous comment, it's adding metadata to packages that has nothing to do with Angular just because it's used by a app using Angular.

**Comment by filipesilva** [comment]
@marcj I do not agree your assessment of how TS is a viable distribution format, and of how npm symlinking should always work because it is an inherent property of npm. But I also don't think it's very productive for me to discuss that here because that's not a necessary part of the issue.

I think the relevant part of this issue is how apps consume their TS sources. That the source is inside node_modules shouldn't be important as long as it's included in the TS compilation.

I checked with the Nx folks and they don't symlink into node_modules, instead they use ts `paths` configuration to direct the app to consume the sources from elsewhere (e.g. `"@my-ws/my-lib": ["libs/my-lib/src/index.ts"]`). This doesn't seem to cause `ngcc` to look at them and try to alter any package.json. 

In your repro you don't use `paths` but instead add those sources via the `includes` array as `"node_modules/@issue1/core/**/*.ts"`. At this point they should be on the TS compilation and compiled with everything else. This seems to cause `ngcc` to look at them and try to alter package.json. 

But ngcc should instead completely ignore those files because they are TS files and the app is compiling them using the Ivy AOT compiler as TS sources.

**Comment by marcj** [comment]
> I do not agree your assessment of how TS is a viable distribution format, and of how npm symlinking should always work because it is an inherent property of npm. But I also don't think it's very productive for me to discuss that here because that's not a necessary part of the issue.

Filipe, what are you trying to pull of here? This and the Angular's opinionated way of implementing Ivy is clearly the root of the problem. You disagree that symlinking is a viable way of linking multiple packages together. Also that TS should not be used directly of a third-party library. This opinion leads to the conclusion that writing to a `package.json` of a package in `node_modules` is legit. Well, if the root assumption is false, then all subsequent conclusions are wrong. Which is clearly the case here.
You assume that modifying a `package.json` is legit. It is not, neither for non-angular packages nor for angular packages. Please stop doing it.
You haven't talked for a second why you even modify foreign files at all. Why is it not a viable option to put your metadata and build artifacts in another place?

The request is pretty clear:

1. Don't write or modify any file in my node_modules because you can't and shouldn't enforce people to not symlink their package into it. By enforcing your opinion you not only limit the functionality of npm itself, but of lerna, yarn workspaces as well. Used by literally ten thousands of projects.
Also this opinion forbids modern full-stack workflows in Lerna, as already described. This will only hurt Angular and force people to use other frameworks for full-stack TS projects.

2. Don't enforce (indirectly or directly) that consumed third-party libraries have to be in JS format.
I mean currently that's not a problem as it works and compiles fine when I consume my TS-only packages and Angular TS-only packages using Ivy. But your communicated opinion can lead pretty quickly to breaking changes regarding this workflow and user requests to maintain the functionality can easily be denied by saying 'it was never officially supported' - as we see impressively here.

> I checked with the Nx folks and they don't symlink into node_modules, instead they use ts paths configuration to direct the app to consume the sources from elsewhere

Ah ok, then they don't use symlinks. Doesn't matter much since Lerna has the big numbers. Literally almost 30.000 public projects rely on symlinked packages alone in Lerna. Native `npm link` and Yarn workspaces is used quite a lot as well. Ignoring their workflow and saying this is no longer supported with Ivy is pretty interesting way of enforcing a bad design decision. I mean, I can understand that in some way. I'm myself a lib author of dozen of libraries and admit to have a design error and fixing it takes some time and motivation, but honest it can't be clearer in this case. Current approach is clearly a dead end.

> But ngcc should instead completely ignore those files because they are TS files and the app is compiling them using the Ivy AOT compiler as TS sources.

It shouldn't matter if the consumed package contains Angular code (and thus something is compiled with `ngtsc` or `ngcc`, I don't know the details) or not. Modifying its source is bad practise as you can see. Just put the artifacts and metadata of those ngts/ngcc commands in your own folder and all of those issues go away.

**Comment by petebacondarwin** [comment]
@marcj thank you for your input on this issue. It is great that we are getting this feedback now. We want to support as many scenarios as we can with ivy. With ngcc we have conflicts of interest since storing its output elsewhere requires tighter integration with build tools. When we first implemented it we decided to first implement a design that would require the least modification to current tooling. We are not fixed with only this approach and as requirements come to light we can try to accommodate them. But as I said there will always be a design tension between the different requirements.

Let's look at ways we can solve the problems presented here.

**Comment by kolkov** [comment]
https://github.com/angular/angular/issues/33469
If I understands correctly, if I enable Ivy in old local library and in my current project, `npm link` will work as before with `build -watch`?

**Comment by petebacondarwin** [comment]
Interestingly (I just checked) if you run ngcc via the command line, without the `-t` option it will traverse the `node_modules` folder but NOT follow any symlinks - so it would miss any `npm link`ed packages.

But... if you use ngcc via CLI integration (or use the command line with `-t`) then it will follow symlinks!! 

This is arguably a bug in the second version, since we really shouldn't be modifying files outside `node_modules`. But then again, as discussed in this issue, one could argue that even modifying files inside node_modules could be deemed rather naughty and not playing nicely with the other kids.

**Comment by marcj** [comment]
> Interestingly (I just checked) if you run ngcc via the command line, without the -t option it will traverse the node_modules folder but NOT follow any symlinks - so it would miss any npm linked packages.

That's because while traversing all packages in `SyncNgccOptions.basePath` symlinks are explicitly excluded when no `targetEntryPointPath` )(`-t`) is given.
https://github.com/angular/angular/blob/168abc6d6f52713383411b14980e104c99bfeef5/packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts#L127-L129

and here an additional redundant check to forbid symlinks:

https://github.com/angular/angular/blob/168abc6d6f52713383411b14980e104c99bfeef5/packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts#L56-L59

I think that this needs either explanation or those checks should be removed. I personally don't see any issues with symlinks. Treating them as hardlinks is probably the best and easiest.

> But... if you use ngcc via CLI integration (or use the command line with -t) then it will follow symlinks!!
> This is arguably a bug in the second version, since we really shouldn't be modifying files outside node_modules.

That depends. Saying `outside node_modules` is not easily identifiable since you can usually disable resolving symlinks using `NODE_PRESERVE_SYMLINKS` or `preserveSymlinks` in `angular.json`. If enabled you usually tell resolvers to not treat symlinks as anything special but just as regular directory and maintain the original path (not the resolved one). So, you no longer know that that code is actually outside of `node_modules`, which is expected behavior. Although it seems that `DirectoryWalkerEntryPointFinder` completely ignores symlinks and always resolves paths, which is something I consider a potential base for further issues.

> one could argue that even modifying files inside node_modules could be deemed rather naughty

:+1:

**Comment by petebacondarwin** [comment]
Sorry not to be more helpful, but just to set expectations.

We are keen to support as many project setups/scenarios as possible with ngcc. But for 9.0.0 projects that use symlinks with their dependencies may not be included in supported ngcc use cases.

Symlinked dependencies are a problem because if we modify the files in there then we could be affecting other projects that depend upon those libraries. One of the original benefits of modifying inside node_modules was that (for most projects) everything in the folder is completely owned by the project containing it. Clearly this is not the case when things are symlinked into it.

For 9.0.0 it may be that disabling ivy is required. We hope to have a solution for this by 9.1.0.

**Comment by JonWallsten** [comment]
@petebacondarwin: You have most certainly already discussed this: but why not create a separate file instead of modifying the package.json? If it's in the node_modules it doesn't matter if it's package.json or random.txt. If it's in our symlinked folders we can easily add it to .gitignore/.npmignore.

**Comment by petebacondarwin** [comment]
@JonWallsten - one reason for modifying the `package.json` is that for the CLI we add new "entry-points", which contain the processed source files. So the changes to this file are not purely markers.

**Comment by PowerKiKi** [comment]
If Ivy does not support symlinked packages, wouldn't it be a deal breaker to use Yarn 2's [PNP feature](url) and all other similar package manager ?

That would be a bad news to have a shiny new compiler that breaks such an interesting new feature from package manager.

**Comment by JonWallsten** [comment]
> If Ivy does not support symlinked packages, wouldn't it be a deal breaker to use Yarn 2's [PNP feature](url) and all other similar package manager ?
> 
> That would be a bad news to have a shiny new compiler that breaks such an interesting new feature from package manager.

It would break that feature anyway since ivy writes metadata in packages inside node_modules and modules are read only in PnP.

**Comment by PowerKiKi** [comment]
Thanks for the confirmation. I was really looking forward to start using PNP and Ivy. I guess that's gonna have to wait :sob:. Hopefully we can find a solution that satisfy everybody.

**Comment by JonWallsten** [comment]
@PowerKiKi Open a new ticket and check what the Angular guys are saying about it at least :)

**Comment by jrood** [comment]
Oof, so if my team wants to adopt a monorepo management system, it sounds like our options are (1) use nx.dev instead of Lerna, (2) use a different frontend framework, or (3) don't upgrade to Ivy. As monorepo management tools gain popularity I wonder if this will become a common trilemma.

**Comment by mgol** [comment]
@jrood You can use Yarn 1, many companies will do so in foreseeable future. In the future, Yarn 2 with the node_modules plugin should work as well.

**Comment by seanlandsman** [comment]
This is a problem if you use Lerna to manage a workspace too I'm afraid

**Comment by marcj** [comment]
@mgol how does Yarn 1 solve this issue? AFAIK they use symlinks as well in Yarn Workspaces. I'm afraid Yarn 2 with the new way to handle node_modules won't be at all compatible with Angular Ivy, but we will see.

**Comment by mgol** [comment]
I understood this issue was about Ivy following symlinks & writing metadata there. That would mean it should work with Yarn 1 & symlinks. But maybe I'm wrong here.

**Comment by marcj** [comment]
@mgol since Yarn 1 uses symlinks as well, their linked packages will get modified package.json as well. So Yarn does not provide a solution here, unfortunately.

**Comment by jrood** [comment]
>I checked with the Nx folks and they don't symlink into node_modules, instead they use ts paths

How can this be? Doesn't Nx work with non-typescript projects? If so, how is typescript a dependency for path resolution?

**Comment by FrozenPandaz** [comment]
I don't think this is pertinent to the conversation here but I'll chime in to clarify 😅.

There are two ways to make the paths work (that I am familiar with and are relevant here):

1. Make the paths actually resolvable with "node module resolution" which most node tools use out of the box by doing some sort of compilation to move the source code into the `node_modules` directory such as linking. This makes it seem like a node_modules that you would install from npm.
  a. This is how lerna and a lot of other tools work.
  b. This way produces side effects to the filesystem which also affects how `ngcc` works.
2. Use path mappings to allow tools to use these mappings to point to _source code_ as a fallback to node module resolution.
  a. This is how Nx and a lot of other tools work :)
  b. This way does not produce side effects to the filesystem but requires tools to accept path mappings or module resolution customization.

You can indeed write your files in `.js` files in Nx. However, the tooling that we provide out of the box uses the second approach above via typescript (which allows those path mappings to be the master set of path mappings for the other tools).

Again, this isn't really pertinent to finding a solution to the issue here.. it just clarifies why Nx, a tool in the ecosystem, is "compatible" with Ivy whereas others currently are not. I don't think other tools should switch to using the second method as they both have tradeoffs. So I digress 😄. If you have more questions about it, we can move the discussion over to the Nx domain to not pollute the discussion here.

**Comment by jrood** [comment]
Thanks for entertaining my tangent @FrozenPandaz !  It's relevant to me because, since Ivy doesn't work with option 1, frankly it puts my team into a place where it looks like we have to either find an alternative to Lerna or find an alternative to Angular.

**Comment by petebacondarwin** [comment]
Hi @jrood - I appreciate your trepidation about this v9 release but hopefully I can offer some glimmers of hope - at least from the ngcc side of things.

First, for the long term, it is important to realise that the aim is for ngcc to be only a temporary requirement. In the not too distant future, we will decide that the new ivy rendering commands are stable enough to allow people to deploy libraries that have been built using ivy. We are not happy to do this just yet because as soon as we allow code compiled with ivy to be published, then the whole set of the ivy rendering instructions effectively become a public API that we are not able to change. Over the next months we will be assessing these instructions and make a call on their stability. Once that happens then we will be pushing to transition the entire ecosystem over to being built and published with ivy enabled. As soon as you have a project that is not using any ViewEngine compiled libraries you will not need ngcc and therefore will not have to worry about how node_modules is managed, since we will not need to modify files in those folders.

Second, for the short term, there is nothing stopping you from disabling ivy in your project, which will allow you to continue with Angular as your framework, while using tooling that expects node_modules to be managed in a way that is currently unsupported by ngcc. The big change in v9 was ivy being turned on by default. But it was always expected that for this release there would still be some projects for which ivy is not yet an option and that is why for v9 ViewEngine compiled projects are still completely supported.

Finally, for the medium term, we are looking seriously at modifications to ngcc and other Angular projects, such as CLI, to allow us to work with tooling that expects node_modules to be managed in a way that is ot currently supported. Scenarios where there are symlinked packages or virtual file-systems do not necessarily allow us unilateral access to modify the files within node_modules. The reason we did not implement such support from day one is that this area of the tooling eco-system is changing daily and it was to risky to attempt to support one approach to find that the tools tack and go in another direction. For example Yarn 2's new plug-n-play has been in active development throughout the time that we have been developing ivy and ngcc.

I would like to lern [sic] more about Lerna's approach to the node_modules folder. I read in their [hoisting](https://github.com/lerna/lerna/blob/master/doc/hoist.md) documentation that symlinking is only an option and only recommended as a workaround for tools that do not follow the node resolution strategy. (ngcc does follow this.)

> Unfortunately, some tooling does not follow the module resolution spec closely, and instead assumes or requires that dependencies are present specifically in the local node_modules directory. To work around this, it is possible to symlink packages from their hoisted top-level location, to individual package node_modules directory. Lerna does not yet do this automatically, and it is recommended instead to work with tool maintainers to migrate to more compatible patterns.

I wonder if it is actually necessary for your project to have this turned on? That being said, since ngcc currently does modify the files in node_modules, it is true that these modification would impact on all packages in your Lerna project, so it might be tricky if you had some packages expecting Ivy and some expecting ViewEngine compiled library code. (This is supported if you are compiling via Angular CLI as we are able to create new ivy entry-points in the library package so that the original ViewEngine code is still available.

---

So the long and the short of it is that there are workarounds in the short-term; that in the long term all will be well; and in the medium term we are keen to try to help as much as possible.

**Comment by akehir** [comment]
> I checked with the Nx folks and they don't symlink into node_modules, instead they use ts `paths` configuration to direct the app to consume the sources from elsewhere (e.g. `"@my-ws/my-lib": ["libs/my-lib/src/index.ts"]`). This doesn't seem to cause `ngcc` to look at them and try to alter any package.json.

Actually, I'm not so sure that works. In my scenario, I have a minimal library and an app to test the library. The issue occurs at least if the library is a compiled Angular library, even though the sources are consumed via typescript path mapping.

If you check out this commit [d882eef
](https://github.com/akehir/fluid-simulation/tree/d882eef7cc5c9cb20fa8ec3802cbe1eefb1689eb) , `npm run ng -- build fluid-simulation-lib --prod` && `npm run test` will  run the tests with Ivy, which modifies the package.json under `"dist/fluid-simulation-lib"`. Therefore the simple workflow of 1) `build library;` 2) `test library as a dependency;` 3) `publish;` breaks. I now either have to compile the tests without ivy, or build the library twice - both of which doesn't seem optimal (I either consume the library with a different compiler than most projects probably will; or I publish different code from the tested code, by building twice).

I appreciate the workarounds available as noted by @petebacondarwin , but I'd say this also highlights another problem with messing with the dependencies; even through typescript path mapping.

**Comment by JonWallsten** [comment]
> Actually, I'm not so sure that works. In my scenario, I have a minimal library and an app to test the library. The issue occurs at least if the library is a compiled Angular library, even though the sources are consumed via typescript path mapping.

Compiling an Angular library with ng-packagr creates a copy of the package.json in the dist folder. Since I started this ticket we have implemented ng-packagr to compile a library, we have also replaced symlinks with path mappings after some guidance from @alxhub.

**Comment by akehir** [comment]
> Compiling an Angular library with ng-packagr creates a copy of the package.json in the dist folder. Since I started this ticket we have implemented ng-packagr to compile a library, we have also replaced symlinks with path mappings after some guidance from @alxhub.
> 
> Edit: Sorry, I read to fast. I see now that you publish in the end. Although if I'm no mistaken Angular recommends you to not publish IVY compiled librarys.

Yes, the library is published in the end; but not built with IVY, but rather VE (`npm run ng -- build fluid-simulation-lib --prod`) . My problem is, that I have an "example-app" project adjacent to the library (in the same angular workspace) which is importing the library as any other app would ( `import { LibraryModule } from @group/library;`). To make this work locally, I use a typescript path mapping from `@group/library` to `dist/library`.

The example project is therefore using the (non-IVY) compiled files from `dist/...` with a typescript path mapping. Running `ng test example-app` with the default configuration will cause the compilation of the example app through IVY, therefore modifying the package.json in the `dist/...` folder.

My point being, running a test of a dependent product / compiling an angular dependency of an angular application will change the code even when using the typescript path mapping.

Reading the issue, it is *expected* behaviour with the current implementation; but I'd still argue it is not expected from a user perspective.

**Comment by JonWallsten** [comment]
@akehir: Ah, now I follow. That one seems tricky. I don't see a "correct" way of doing that.
Build lib with Ivy, no changed package.json, but you shouldn't publish. 
Build with VE, can't use published version because it's not published yet. Duh.
Build with VE and use symlink/path mapping, will change package.json, but you can publish.
If I were you I would probably create a small script that removes the code from the package.json file and run that code in the prepublish hook. Although I think that Angular CLI would offer this code through some ng <whatever> script.

**Comment by petebacondarwin** [comment]
@akehir - this is a scenario that I have seen work before - unless I am misunderstanding...
The library can be built with ViewEngine and ngPackagr, which is how it will also be distributed. As part of your build you can then run `npm pack` to generate a `.tgz` file of the library.
Then in your test app you just point to this `.tgz` as a dependency and a new copy will be created in the test application node_modules folder, which ngcc can safely compile to ivy for application testing, while the original dist folder is untouched.

**Comment by akehir** [comment]
> Then in your test app you just point to this `.tgz` as a dependency and a new copy will be created in the test application node_modules folder, which ngcc can safely compile to ivy for application testing, while the original dist folder is untouched.

But this approach means that I'll have to abandon the approach of having a self-contained *monorepo* for both library and test application. Currently, the design is quite elegant of having library and test application as adjacent applications in the same angular workspace.

With npm pack, I'd have to have the angular 'workspace' of the library contain a reference to itself and run an npm install after every build of the library. The path mapping resolves the dependency in a transparent way.

@JonWallsten : Either editing the package.json or just creating a copy of the /dist folder would work easily. For now, my solution is just using VE for building the tests. For me it was just very surprising that running the tests was changing my build artefacts in the test folder.

**Comment by petebacondarwin** [comment]
Just to be clear @akehir - have you discounted building the example app directly against the TS sources of the lib. E.g. changing these lines: https://github.com/akehir/fluid-simulation/blob/d882eef7cc5c9cb20fa8ec3802cbe1eefb1689eb/tsconfig.json#L21-L27

To be something like:

```
    "paths": {
      "@triangular/fluid-simulation": [
        "projects/fluid-simulation-lib/src/lib"
      ],
      "@triangular/fluid-simulation/*": [
        "projects/fluid-simulation-lib/src/lib/*"
      ]
```

This would build the library directly with ivy just for your example app. You would then be testing the ivy version of the code. When you deploy, although the code is built with ViewEngine, ngcc will also convert the code to ivy, so your downstream clients will be effectively using the same ivy generated code as you test in your example app....

**Comment by akehir** [comment]
@petebacondarwin Thanks for the input. Changing the configuration as per your suggestion also works and keeps the package.json intact. My intuition was that I desired to test the packaged / built version of the library against the downstream example app. The idea was to confirm that there's nothing with the build step; and with what I was planning to npm publish.

**Comment by petebacondarwin** [comment]
@akehir - perhaps you have a pre-publish sanity check test that does actually pack the library into a `.tgz` file and runs example app tests against that? That way you get fast turnaround when developing but you have the confidence of having tested the "real thing" before publishing. Actually depending upon your audience you might also want to test an app in ViewEngine mode using the library too.

**Comment by Nightbr** [comment]
Hey,

we use lerna & Angular 9 ivy and we have the same issue. The problem is it breaks the production build for us which is more critical.

We have a lerna monorepo with:

```
root
- frontend
  - app-panel (Angular 9 Ivy project)
  - design-system
     - angular (angular adapter for design system build with ng-packagr)
     - core (base design system component build with stenciljs)
```

`AppPanel` has `design-system/angular` as a local dependency & lerna symlink the `dist` folder into AppPanel node_modules. 

When we run `ng serve` or `ng serve --prod` (same behavior with build) the first time we have a weird error:

```
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'selectors' of null
TypeError: Cannot read property 'selectors' of null
    at main-es2015.ada6b062a2963d52a497.js:1
    at Ms (main-es2015.ada6b062a2963d52a497.js:1)
    at main-es2015.ada6b062a2963d52a497.js:1
    at Object.ya (main-es2015.ada6b062a2963d52a497.js:1)
    at template (77-es2015.cbafd8da93c415dffc75.js:1)
    at ks (main-es2015.ada6b062a2963d52a497.js:1)
    at Ts (main-es2015.ada6b062a2963d52a497.js:1)
    at ti (main-es2015.ada6b062a2963d52a497.js:1)
    at main-es2015.ada6b062a2963d52a497.js:1
    at Ts (main-es2015.ada6b062a2963d52a497.js:1)
    at T (polyfills-es2015.e94d0446fd672132aed7.js:1)
    at T (polyfills-es2015.e94d0446fd672132aed7.js:1)
    at polyfills-es2015.e94d0446fd672132aed7.js:1
    at l.invokeTask (polyfills-es2015.e94d0446fd672132aed7.js:1)
    at Object.onInvokeTask (main-es2015.ada6b062a2963d52a497.js:1)
    at l.invokeTask (polyfills-es2015.e94d0446fd672132aed7.js:1)
    at i.runTask (polyfills-es2015.e94d0446fd672132aed7.js:1)
    at m (polyfills-es2015.e94d0446fd672132aed7.js:1)
```

When we kill and run a second it works.

Seems "the symlink content need to be refreshed" because Ivy writes `__ivy_ngcc__` when compiling AppPanel into `design-system/angular` package in node_modules.

`Compiling @myorga/ds-angular : es2015 as esm2015`

We tried to build the Design System with Ivy Enable -> Same error

Add to that we have `"preserveSymlinks": true,` properly enable in our angular.json of AppPanel.

For development, we just start twice and it works on the second time but it's for our build in CI which is more critical, we can't build twice...

**WORKAROUND for our CI build**
We unlink the symlink after lerna bootstrap & hard copy the dependency...

```
cd node_modules/@myorga
unlink ds-angular
cd ../../../design-system/angular
cp -a dist/. ../../app-panel/node_modules/@myorga/ds-angular
cd ../../app-panel
```

**Comment by Nightbr** [comment]
Add to that, here is the error in dev mode (first compilation):

```
Uncaught Error: Angular JIT compilation failed: '@angular/compiler' not loaded!
  - JIT compilation is discouraged for production use-cases! Consider AOT mode instead.
  - Did you bootstrap using '@angular/platform-browser-dynamic' or '@angular/platform-server'?
  - Alternatively provide the compiler with 'import "@angular/compiler";' before bootstrapping.
    at getCompilerFacade (vendor.js:20076)
    at Function.get (vendor.js:53355)
    at registerNgModuleType (vendor.js:51017)
    at vendor.js:51035
    at Array.forEach (<anonymous>)
    at registerNgModuleType (vendor.js:51031)
    at new NgModuleFactory$1 (vendor.js:51183)
    at compileNgModuleFactory__POST_R3__ (vendor.js:56111)
    at PlatformRef.bootstrapModule (vendor.js:56440)
    at Module../src/main.ts (main.js:1493)
```

the second compilation works fine (no error).

**Comment by CDevine1** [comment]
I saw a workaround posted at  #35447 which basically has you run ngcc ahead so that it won’t run at compile time.

In case it helps:
> Add the `"postinstall"` documented in https://angular.io/guide/ivy#speeding-up-ngcc-compilation, then `rm -rf node_modules`, then `yarn` if using yarn workspaces (or `lerna` in your case). This fixed it for me, our project is using yarn workspaces instead of lerna.
> 
> If it fixes it for the others too, the guide's title could be changed to "Ngcc compilation with lerna, yarn workspaces, or just speeding up ngcc compilation in general.

**Comment by Nightbr** [comment]
It doesn't work with lerna. 

In lerna, if you do `lerna bootstrap` to `install`, `prepare`, `symlink` your packages together. 

If you do on `postinstall`, it will run the `ngcc` after the install so it will not take into account other angular packages so we get the same error & need to build twice.

I tried to put `ngcc` in the `prepare`:
```json
{
    "prepare": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points",
}
```
It will compile properly all libs installed but not other angular packages of the monorepo because they have not been symlinked yet... So same error, need to run angular app twice to work...

I'm looking for a `postBootstrap` lifecycle in lerna but it seems it doesn't exist...

**Comment by Lonli-Lokli** [comment]
> For 9.0.0 it may be that disabling ivy is required. We hope to have a solution for this by 9.1.0.

Seems like it was not fixed?
Is there any plans to fix it in the next 9.x or at least 10.0 ?

**Comment by petebacondarwin** [comment]
This is proving to be difficult to resolve. We still don't have a solution for symlinked packages. It is not even clear if we can find a resolution for v10.

**Comment by petebacondarwin** [comment]
We have two related issues here:

1) **ngcc doesn't work well with sym-linked libraries** - in particular we need a better story for library authors to build applications that test the libraries.
2) ngcc doesn't work well with the standard lerna setup - it seems possible that there is a [workaround for lerna](https://github.com/angular/angular/issues/33395#issuecomment-603113739) that will allow ngcc to process the necessary packages but no one on the Angular team has enough experience or knowledge to understand if this is valid.

**Comment by mgol** [comment]
About the Lerna workaround: Lerna only does this package.json hacky rewriting if it needs to manage workspaces by working around the package manager restrictions. I don’t recommend using Lerna like that, it only causes pain & bugs in edge cases like this one. If you use Lerna with Yarn by relying on Yarn internal workspaces support instead (there’s a Lerna setting for that) all this hacky behavior is gone as Lerna leaves managing workspaces to Yarn & it doesn’t rewrite package.json.

**Comment by Nightbr** [comment]
@mgol ok if you use `yarn` but if you use `npm` you have to stick with the "hacky" rewriting. So still need to handle that case I think.

**Comment by mgol** [comment]
@Nightbr Yes, sure! I just don’t consider npm to be suitable package manager to use with Lerna because of missing support for workspaces.

**Comment by mischkl** [comment]
We have the same problem, and there's another wrinkle to it: We have a shared library that's used by two different applications, one using Angular v8 (View Engine) and one using Angular v10 (Ivy). Up until now we relied on symlinking the library into both applications - but now we can't because as soon as the Angular 10 application is run, it modifies the shared library and thus the Angular v8 library can't deal with it any more.

It seems to me that modifying the contents of anything dependencies is a fundamental design flaw of the Ivy compile infrastructure, that makes it completely incompatible with symlinking, which a lot of developers depend on. In my mind it shouldn't be too hard to simply use a cache folder for the Ivy-compiled versions of View Engine packages, instead of rewriting those packages' contents.

**Comment by dennisameling** [comment]
Ran into the same issue with https://github.com/dennisameling/muuri-angular and the workaround below got me up and running at least (basically an implementation of what @petebacondarwin mentioned in https://github.com/angular/angular/issues/33395#issuecomment-546427262). These steps are also described at https://github.com/dennisameling/muuri-angular#contributing.

**Update library config**
With the steps below, we'll add some additional configuration to your library so that it can more easily be consumed by projects that consume your library using `npm link`.
- In `projects/PROJECT_NAME`, it's very likely you already have a `tsconfig.lib.prod.json` because `ng generate library` already [defaults to the View Engine](https://angular.io/guide/creating-libraries#publishing-your-library) for publishing to NPM.
- Create a new `tsconfig.lib.ivy.json` in `projects/PROJECT_NAME` with the following contents:
```JSON
/* To learn more about this file see: https://angular.io/config/tsconfig. */
{
  "extends": "./tsconfig.lib.json",
  "angularCompilerOptions": {
    "enableIvy": true
  }
}
```
- In your library's `angular.json` file, under `projects > PROJECT_NAME > architect > build`, add a new configuration `development` so it looks like this:
```JSON
          "configurations": {
            "production": {
              "tsConfig": "projects/PROJECT_NAME/tsconfig.lib.prod.json"
            },
            "ivy": {
              "tsConfig": "projects/PROJECT_NAME/tsconfig.lib.ivy.json"
            }
          }
```
- Update your `package.json` for easy access to the following commands:
```JSON
  "scripts": {
    ...
    "build:prod": "ng build PROJECT_NAME --prod",
    "build:ivy": "ng build PROJECT_NAME --configuration=ivy"
  },
```

**For Angular 9+ (Ivy) projects consuming your library**
- Run `npm run build:ivy`. This will build an Ivy-compatible library that you can use in Angular 9+ (Ivy) projects locally.
- Run `cd dist/PROJECT_NAME`.
- Run `npm link`.
- In your project, run `npm link PROJECT_NAME` to link to your library. Your project will now use the local `PROJECT_NAME` code.
- Bonus tip: Run `npm run build:ivy -- --watch` so that the library gets rebuilt on every code change 😃

**For Angular 8 and lower (View Engine) projects consuming your library**
- Run `npm run build:prod`. This will build the library with the legacy View Engine so that it can easily be consumed by non-Ivy projects. This would also be the build that you publish on NPM.
- Run the same steps as above for Ivy projects.

**Comment by Lonli-Lokli** [comment]
Seems like there are plans to fix this issue in future versions
https://github.com/angular/angular/issues/38366

**Comment by elclanrs** [comment]
I'm having the same issue with pnpm. Running `ng serve` a second time works as noted in previous posts. I am able to resolve this issue for now by manually compiling all node_modules in postinstall:

```json
"postinstall": "find node_modules/{@angular,ngx-mask,ng-dynamic-component} | xargs -I% pnpx ngcc -s %"
```

`ngx-mask` and `ng-dynamic-component` are Angular packages that are needed to run the app.

Note that `ngcc` alone did not work and I had to explicitly run it on all modules. The downside is that it compiles _all_ Angular modules, even ones I don't use or need. I suppose I could be more specific in what to compile, but it would require more maintenance.

**Comment by petebacondarwin** [comment]
This should be solved by the ng-linker that is being developed. Please follow the progress of that here: https://github.com/orgs/angular/projects/2
